### DISCUSSION TITLE
Cannot derive Parser

### DISCUSSION THREAD

**Discussion by Proximyst** [pest-parser/pest#648] [General]
With `src/pest.rs` having:
```rust
use pest_derive::*;

#[derive(Parser)]
#[grammar = "grammar.pest"]
pub struct Parser;
```

on `pest_derive v2.1.0` and `pest v2.1.0`/`pest v2.1.2` (same error on both versions), it returns the following errors:

```
error[E0433]: failed to resolve: could not find `iterators` in `pest`
 --> src/pest.rs:4:10
  |
4 | #[derive(Parser)]
  |          ^^^^^^ could not find `iterators` in `pest`

error[E0433]: failed to resolve: could not find `error` in `pest`
 --> src/pest.rs:4:10
  |
4 | #[derive(Parser)]
  |          ^^^^^^ could not find `error` in `pest`

error[E0433]: failed to resolve: could not find `Atomicity` in `pest`
 --> src/pest.rs:4:10
  |
4 | #[derive(Parser)]
  |          ^^^^^^ could not find `Atomicity` in `pest`

error[E0433]: failed to resolve: could not find `unicode` in `pest`
 --> src/pest.rs:4:10
  |
4 | #[derive(Parser)]
  |          ^^^^^^ could not find `unicode` in `pest`

error[E0404]: expected trait, found struct `pest::Parser`
 --> src/pest.rs:4:10
  |
4 | #[derive(Parser)]
  |          ^^^^^^ not a trait

error[E0412]: cannot find type `ParserState` in module `pest`
 --> src/pest.rs:4:10
  |
4 | #[derive(Parser)]
  |          ^^^^^^ not found in `pest`

error[E0412]: cannot find type `ParseResult` in module `pest`
 --> src/pest.rs:4:10
  |
4 | #[derive(Parser)]
  |          ^^^^^^ not found in `pest`

error[E0425]: cannot find function `state` in module `pest`
 --> src/pest.rs:4:10
  |
4 | #[derive(Parser)]
  |          ^^^^^^ not found in `pest`

```

This miiight be related to #129

**Comment by lwandrebeck** [comment]
adding `use pest::Parser;` should do the trick. Plus I won’t use `pub struct Parser;` but another name for the struct.
HTH,

**Comment by dragostis** [comment]
Does the `Parser` name break this so terribly? If so, it makes sense to have a better error in pest_derive.

**Comment by CAD97** [comment]
What's happening here is that the derive is trying to find its items in your `mod pest`. If you use a different module name, it should work. (I've used `mod lang { #[derive(pest::Parser)] struct Parser; }` many times with no issue.)

@Proximyst what edition are you compiling in? This should not be able to happen in edition=2018. Due to the way proc-macros work, pest requires that the path `::pest` refers to the pest crate.

**Comment by Proximyst** [comment]
Your suggestion works, @CAD97. I'm using 2018 edition; rustc 1.40.0-nightly (87cbf0a54 2019-11-01).

**Comment by monoclex** [comment]
Any word on this? This should be solvable by prepending `::` into whatever macro generates the code, so that it looks like `::pest::stuff`.

Here's another workaround, if you're in a `pest.rs` module:
```rust
pub use pest::{error, iterators, state, ParseResult, ParserState};
```

**Comment by berlinshawn063** [comment]
I discovered this was for 2 reasons: 1) there was a typo in the .pest file that made the proc crash causing a compiler error; and 2) you need `#[derive(pest_derive::Parser)] to [Hy-Vee Consumer Survey](https://graspsurvey.com/hyvee-survey/) it from pest::Parser.
